guides/source/action_controller_advanced_topics.md

```diff
@@ -0,0 +1,500 @@
+
+### RESTful Downloads
+
+While `send_data` works fine, if you are creating a RESTful application having separate actions for file downloads is usually not necessary. In REST terminology, the PDF file from the example above can be considered just another representation of the client resource. Rails provides a slick way of doing "RESTful" downloads. Here's how you can rewrite the example so that the PDF download is a part of the `show` action, without any streaming:
+
+```ruby
+class ClientsController < ApplicationController
+  # The user can request to receive this resource as HTML or PDF.
+  def show
+    @client = Client.find(params[:id])
+
+    respond_to do |format|
+      format.html
+      format.pdf { render pdf: generate_pdf(@client) }
+    end
+  end
+end
+```
+
+Now the user can request to get a PDF version of a client just by adding ".pdf" to the URL:
+
+```
+GET /clients/1.pdf
+```
+
+You can call any method on `format` that is an extension registered as a MIME type by Rails. Rails already registers common MIME types like `"text/html"` and `"application/pdf"`:
+
+```ruby
+Mime::Type.lookup_by_extension(:pdf)
+# => "application/pdf"
+```
+
+If you need additional MIME types, call [`Mime::Type.register`](https://api.rubyonrails.org/classes/Mime/Type.html#method-c-register) in the file `config/initializers/mime_types.rb`. For example, this is how you would register the Rich Text Format (RTF):
+
+```ruby
+Mime::Type.register("application/rtf", :rtf)
+```
+
+NOTE: If you modify an initializer file, you have to restart the server for their changes to take effect.
+
+### Live Streaming of Arbitrary Data
+
+Rails allows you to stream more than just files. In fact, you can stream
+anything you would like in a response object. The
+[`ActionController::Live`](https://api.rubyonrails.org/classes/ActionController/Live.html)
+module allows you to create a persistent connection with a browser. By including
+this module in your controller, you can send arbitrary data to the browser at
+specific points in time.
+
+```ruby
+class MyController < ActionController::Base
+  include ActionController::Live
+
+  def stream
+    response.headers["Content-Type"] = "text/event-stream"
+    100.times {
+      response.stream.write "hello world\n"
+      sleep 1
+    }
+  ensure
+    response.stream.close
+  end
+end
+```
+
+The above example will keep a persistent connection with the browser and send
+100 messages of `"hello world\n"`, each one second apart.
+
+Note that you have to make sure to close the response stream, otherwise the
+stream will leave the socket open indefinitely. You also have to set the content
+type to `text/event-stream` *before* calling `write` on the response stream.
+Headers cannot be written after the response has been committed (when
+`response.committed?` returns a truthy value) with either `write` or `commit`.
+
+#### Example Use Case
+
+Let's suppose that you were making a Karaoke machine, and a user wants to get
```

Comment:
I don't think "karaoke" should be titlecase.

Reply:
sure